src/main/kotlin/com/github/codefabrikgmbh/scenarios/Given.kt

```diff
@@ -16,36 +16,36 @@ class ScenarioRunner<T : Scenario>(
     private val executeSteps: T.() -> Unit
 ) {
     fun run() {
-        run<Unit>(null)
+        runExpecting<Unit>(null)
     }
 
     @Suppress("UNCHECKED_CAST")
-    fun <K : Any> run(expected: KClass<K>?, runSteps: T.(K) -> Unit = {}) {
+    fun <K : Any> runExpecting(expected: KClass<K>?, expectationSteps: T.(K) -> Unit = {}) {
         val scenarioResult = runCatching(scenarioSupplier)
         scenarioResult.onFailure {
             afterScenario()
             throw AssertionError("Context is invalid.")
         }
 
-        val stepsResult = scenarioResult.mapCatching(this.executeSteps)
+        val executionResult = scenarioResult.mapCatching(this.executeSteps)
 
         if (expected == null) {
             afterScenario()
-            stepsResult.onFailure { throw it }
+            executionResult.onFailure { throw it }
         } else {
-            val exception = stepsResult.exceptionOrNull() ?: throw AssertionError("Expected error $expected has not been thrown.")
-            val verificationResult = scenarioResult.mapCatching { runSteps(it, exception as K) }
+            val exception = executionResult.exceptionOrNull() ?: throw AssertionError("Expected error $expected has not been thrown.")
+            val expectationResult = scenarioResult.mapCatching { expectationSteps(it, exception as K) }
 
             afterScenario()
 
-            stepsResult.onSuccess {
+            executionResult.onSuccess {
                 throw AssertionError("Expected error $expected has not been thrown.")
             }
-            stepsResult.onFailure {
-                if (it::class == expected) null else throw it
+            executionResult.onFailure {
```

Comment:
This should happen before the expectation steps are executed, otherwise the `exception as K` cast may fail.